Movements outside grab range stops grab

HMBGERDO opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
<!-- Include a small to medium description of what your PR changes. -->
<!-- Document all changes, as not doing this may delay reviews or even discourage maintainers from merging your PR! -->
<!-- If your PR fixes an issue, add "Fixes #1234" somewhere in the PR description. This will automatically close the bug upon PR submission. -->
This PR adds check for target movement with signals so if target was moved and it's not adjucent to us, grab will fail

## Why It's Good For The Game
<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Its preventing table fu, Charlie wanted this realisation in signals

## Testing
<!-- How did you test the PR, if at all? -->
Compiled, grabbed

## Changelog
:cl:
tweak: Grab now fails at the moment target is not adjacent to user.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

HMBGERDO commented:
i will reopen when i end refactoring this shit

HMBGERDO commented:
there is still an issue with affected animation when you play with diagonal movement but i dunno what to do about it
https://youtu.be/3KCGjDxvIaE

HMBGERDO commented:
dual grab was changed a bit, sometimes your victims switch positions but should be fine
https://youtu.be/Zh9u1OKa1ww

Komrad822 commented:
Ok, you dont like table fu. Why

DGamerL commented:
> Ok, you dont like table fu. Why

It is quite literally a bug in the game, which also causes things like just pulling someone out of a grab can cause them to teleport back if the grab is reinforce.

S34NW commented:
(PR is approved by balance team, but labels mean it's not properly showing)

farie82 commented on code/modules/mob/mob_grab.dm:
```diff
@@ -62,6 +64,55 @@
+	var/possible_dest = list(old_turf)
+	for(var/turf/dest in orange(assailant, 1))
+		if(dest.Adjacent(affecting))
+			possible_dest |= dest
```
```suggestion
			possible_dest += dest
```
You don't need to use `|=`, since `orange` only returns a turf once

farie82 replied:
Oh yep you're right ofc. I didnt think of the fact that old_turf exists

farie82 commented on code/modules/mob/mob_grab.dm:
```diff
@@ -62,6 +64,55 @@
+	if(istype(assailant.r_hand, /obj/item/grab))
+		var/obj/item/grab/grab = assailant.r_hand
+		possible_dest -= get_turf(grab.affecting)
+	for(var/turf/dest in possible_dest)
```
```suggestion
	for(var/turf/dest as anything in possible_dest)
```
It only has turfs in the list. So `as anything` will make it so that Byond won't apply a filter on the turf type, slightly faster

farie82 commented on code/modules/mob/mob_grab.dm:
```diff
@@ -62,6 +64,55 @@
+		if(affecting.Move(dest, get_dir(affecting, dest), assailant.movement_delay() + 7))
+			break
+	if(state == GRAB_NECK)
+		assailant.setDir(angle2dir((dir2angle(direct) + 202.5) % 365))
```
```suggestion
		assailant.setDir(angle2dir((dir2angle(direct) + 202.5) % 360))
```
365 is the amount of days per year, not the amount of degrees in a full circle :P
Oh I just noticed this is the old code lol

HMBGERDO replied:
yep, this part of code is shit and it was not working at all, i missed it(and seems like other guys too since there is no issues)
this change wont make it work, gonna commit fix soon

farie82 reviewed (commented):
Few minor things

HMBGERDO commented:
okay uh, it was rebased on master, issue with facing back neck grab is fixed, we should be fine now

Qwertytoforty commented on code/game/atoms_movable.dm:
```diff
@@ -163,6 +163,7 @@
 /atom/movable/Move(atom/newloc, direct = 0, movetime)
 	if(!loc || !newloc) return 0
 	var/atom/oldloc = loc
+	setDir(direct)
```
What is the purpose of this line? Breaks things like directional lock, and I am curious what other effects it can have, since this is *all* moveable objects, rather than just carbons

HMBGERDO replied:
This was in client/Move after Move proc called, because of that you were looking at the direction you're going
Firstly, i removed it because i thought inbuilt proc move that is called when we want to move something changes dir to the move direction and it worked fine until i started crawling. While crawling without this thing you were able to crawl with legs infront which is weird so i placed this thing here back. 
This should not break directional lock since directional lock is overrided in SetDir proc, but i will test it and find out the problem
I probably gonna place it in carbon/Move proc instead of this

HMBGERDO replied:
oh, i forgot to mention, i removed it from client/Move because it was breaking setting dir in Move proc, meaning when you hold neck grab and move, you will be in same direction as if you move in any other situations, that is why i removed it.

HMBGERDO replied:
fuck, i found why this is breaking forced look. Inbuild proc Move changes your direction as i said above and if you set dir before proc Move, inbuild proc overrides character dir
bit if you set dir after move, it means you cant change dir in Move proc like we change currently with neck grab

HMBGERDO replied:
the problem is fixed and the proc is moved to carbon/human in last update

HMBGERDO commented:
okay so, last updates did
1: fixed directional lock not working, works again as intended
2: fixed flipping with grab not swapping positions
3: fixed kill and neck grab animation speed being different from assailent moving animation speed

DGamerL commented on code/game/atoms_movable.dm:
```diff
@@ -169,6 +169,7 @@
 			glide_for(movetime)
 		if(!(direct & (direct - 1))) //Cardinal move
 			. = ..(newloc, direct) // don't pass up movetime
+			setDir(direct)
```
Why do you do this every `Move`? This seems quite hacky

DGamerL replied:
Alright fair enough!

DGamerL commented on code/modules/mob/mob_grab.dm:
```diff
@@ -62,6 +64,57 @@
+	if(!assailant.Adjacent(old_turf))
+		qdel(src)
+		return
+	var/grab_states_not_moving = list(GRAB_KILL, GRAB_NECK) //states of grab when we dont need affecting to be moved by himself
```
```suggestion
	var/list/grab_states_not_moving = list(GRAB_KILL, GRAB_NECK) //states of grab when we dont need affecting to be moved by himself
```

DGamerL commented on code/modules/mob/mob_grab.dm:
```diff
@@ -62,6 +64,57 @@
+		qdel(src)
+		return
+	var/grab_states_not_moving = list(GRAB_KILL, GRAB_NECK) //states of grab when we dont need affecting to be moved by himself
+	var/assailant_glide_speed = TICKS2DS(world.icon_size/assailant.glide_size)
```
```suggestion
	var/assailant_glide_speed = TICKS2DS(world.icon_size / assailant.glide_size)
```

DGamerL commented on code/modules/mob/mob_grab.dm:
```diff
@@ -169,9 +222,9 @@
-	if(IS_HORIZONTAL(affecting) && state != GRAB_KILL)
-		animate(affecting, pixel_x = 0, pixel_y = 0, 5, 1, LINEAR_EASING)
-		return //KJK
+	// if(IS_HORIZONTAL(affecting) && state != GRAB_KILL)
+		// animate(affecting, pixel_x = 0, pixel_y = 0, 5, 1, LINEAR_EASING)
+		// return //KJK
```
If we don't use this code anymore, could you remove this?

DGamerL reviewed (commented):
Code looks good, except for that `setDir`, apologies for the long radiosilence!

Contrabang commented:
Sorry for this being in limbo for so long, its just like... no one wants to touch this code.

The code looks technically fine, but upon testing, you get a lot of weird stuff like this. I think one of the major errors is that its not deterministic in the way that they move, instead picking a random valid tile.

https://github.com/ParadiseSS13/Paradise/assets/91113370/7145b3c7-42ce-4092-b47d-af2e83252e56

(It looks like the game lagged in the middle and I teleported 3 blocks but nope, thats actually just the grabs being fucked up)

https://github.com/ParadiseSS13/Paradise/assets/91113370/53288771-d832-4d38-9d78-b97b3d861942

(23 seconds in on this 3rd video)

https://github.com/ParadiseSS13/Paradise/assets/91113370/55458b09-d60f-4ddf-8109-526362a11707

DGamerL commented:
> okay so teleports are gone(or at least should be gone(but i really hope they are gone forever)) diagonal movement bugs(which also called teleport, which called loosing grab when diagonal moving into the wall) are gone
> 
> !!! added Moved() call in Move() proc after we did first diagonal step and failed to make second one i dont think this would break something but can we TM this again if the code is fine

Did you test the latest commits and didn't find anything like shown in Contra's videos? If so, I'll ask if this can be TMed again!

HMBGERDO commented:
> > okay so teleports are gone(or at least should be gone(but i really hope they are gone forever)) diagonal movement bugs(which also called teleport, which called loosing grab when diagonal moving into the wall) are gone
> > 
> > !!! added Moved() call in Move() proc after we did first diagonal step and failed to make second one i dont think this would break something but can we TM this again if the code is fine
> 
> Did you test the latest commits and didn't find anything like shown in Contra's videos? If so, I'll ask if this can be TMed again!

The last commit fixes 2/3 of issues in contra's videos
The last one can't really be fixed or i don't know yet how to fix that, that's one with guys swapping positions when you move

S34NW reviewed (approved):
Two weeks without issue on TM
